lightning-net-tokio/src/lib.rs

```diff
@@ -1,279 +1,640 @@
-						let mut lock = us_ref.lock().unwrap();
-						lock.read_paused = true;
+
+			macro_rules! prepare_read_write_call {
+				() => { {
+					let mut us_lock = us.lock().unwrap();
+					if us_lock.disconnect_state == DisconnectionState::RLTriggeredDisconnect {
+						shutdown_socket!("disconnect_socket() call from RL");
 					}
-				},
-				Err(e) => {
-					us_ref.lock().unwrap().need_disconnect = false;
-					return future::Either::B(future::result(Err(std::io::Error::new(std::io::ErrorKind::InvalidData, e))));
-				}
+					us_lock.block_disconnect_socket = true;
+				} }
 			}
 
-			if let Err(e) = us_ref.lock().unwrap().event_notify.try_send(()) {
-				// Ignore full errors as we just need them to poll after this point, so if the user
-				// hasn't received the last send yet, it doesn't matter.
-				assert!(e.is_full());
+			// Whenever we want to block on reading or waiting for reading to resume, we have to
+			// at least select with the write_avail_receiver, which is used by the
+			// SocketDescriptor to wake us up if we need to shut down the socket or if we need
+			// to generate a write_buffer_space_avail call.
+			macro_rules! select_write_ev {
+				($v: expr) => { {
+					assert!($v.is_some()); // We can't have dropped the sending end, its in the us Arc!
+					prepare_read_write_call!();
+					if let Err(e) = peer_manager.write_buffer_space_avail(&mut SocketDescriptor::new(us.clone())) {
+						us.lock().unwrap().disconnect_state = DisconnectionState::RLTriggeredDisconnect;
+						shutdown_socket!(e);
+					}
+					us.lock().unwrap().block_disconnect_socket = false;
+				} }
 			}
 
-			future::Either::B(future::result(Ok(())))
-		}).then(move |_| {
-			if us_close_ref.lock().unwrap().need_disconnect {
-				peer_manager_ref.socket_disconnected(&SocketDescriptor::new(us_close_ref, peer_manager_ref.clone()));
-				println!("Peer disconnected!");
-			} else {
-				println!("We disconnected peer!");
+			tokio::select! {
+				v = write_avail_receiver.recv() => select_write_ev!(v),
+				read = reader.read(&mut buf) => match read {
+					Ok(0) => {
+						println!("Connection closed");
+						break;
+					},
+					Ok(len) => {
+						if let Some(blocker) = {
+							let mut lock = us.lock().unwrap();
+							if lock.disconnect_state == DisconnectionState::RLTriggeredDisconnect {
+								shutdown_socket!("disconnect_socket() call from RL");
+							}
+							if lock.read_paused {
+								let (sender, blocker) = oneshot::channel();
+								lock.read_blocker = Some(sender);
+								Some(blocker)
+							} else { None }
+						} {
+							tokio::select! {
+								res = blocker => {
+									res.unwrap(); // We should never drop the sender without sending () into it!
+									if us.lock().unwrap().disconnect_state == DisconnectionState::RLTriggeredDisconnect {
+										shutdown_socket!("disconnect_socket() call from RL");
+									}
+								},
+								v = write_avail_receiver.recv() => select_write_ev!(v),
+							}
+						}
```

Comment:
The logic around paused reads with the nested `select!` is a bit difficult to follow. I wonder if this could be simplified a bit using a precondition on the read branch.

```rust
let read_paused = us.lock().unwrap().read_paused;
tokio::select! {
    v = write_avail_receiver.recv() => {
        // ...
    },
    read = reader.read(&mut buf), if !read_paused => match read {
        // ...
    },
}
```

Then I think this would eliminate the need for the blocker and thus the nested `select!`. Instead, we would write until `send_data` indicates to resume reading, in which case the read branch would be reenabled.

Reply:
Hmm, I'm not convinced that select!{} definitely wont call the read() future if !read_paused (instead of blocking on the future and never taking that branch). I'd buy that it likely wont, but there's sadly way too little documentation on it.

Reply:
In reading the documentation, I realize my comment was definitely not right, as the docs explicitly call out this case and say that the future wont be polled. That said, I don't think it eliminates the need for the blocker - we still need a way to wake up the reader in case we got a write with resume_read set.

I initially tried to rewrite it with just one Sender which could wake the read task, but sadly we need to make sure that if we wake it due to a write_avail event, we always call write_buff_space_avail. This would mean either the channel has an unlimited depth and we scan it for any write_avail events, or we add yet another bool to indicate that write_avail needs to be called. I find both worse than keeping the current model of just having two Senders, but I removed the Option<> and simplified the logic a ton.

Reply:
Well, it *should* call the future returned by `read()` if `!read_paused` is true. I guess you meant if `!read_paused` is false? Maybe I'm not parsing your comment correctly.

Anyhow, I found this documentation, which is fairly extensive and indicates that `select!` behaves as you would expect (i.e., an `await` won't occur on a disabled branch).

https://docs.rs/tokio/0.2.13/tokio/macro.select.html

Any ambiguity is easily testable with the following code, if I understand your concern correctly:

```rust
async fn do_panic() {
    unreachable!();
}

tokio::select! {
    _ = do_panic(), if false => unreachable!(),
    else => assert!(true),
}
```

Reply:
Didn't notice your reply before commenting, so deleted it. Looks much simpler now!

Reply:
Ah, with regard to the blocker, isn't writing the only way to resume reading? That is, we can't resume reading until `send_data` is called with `resume_read` set `true`. That was why I thought the blocker was not needed. I haven't fully reviewed the `send_data` implementation, so I could be missing something.